metered-channel/src/bounded.rs

```diff
@@ -164,39 +174,41 @@ impl<T> MeteredSender<T> {
 	}
 
 	/// Send message, wait until capacity is available.
-	pub async fn send(&mut self, msg: T) -> result::Result<(), SendError>
+	pub async fn send(&mut self, msg: T) -> result::Result<(), SendError<T>>
 	where
 		Self: Unpin,
 	{
 		match self.try_send(msg) {
 			Err(send_err) => {
 				if !send_err.is_full() {
-					return Err(SendError::Disconnected)
+					return Err(SendError::Disconnected(send_err.into_inner().into()))
 				}
 
-				let msg = send_err.into_inner();
+				let msg = send_err.into_inner().into();
 				self.meter.note_sent();
 				let fut = self.inner.send(msg);
 				futures::pin_mut!(fut);
-				fut.await.map_err(|_| {
+				fut.await.map_err(|err| {
 					self.meter.retract_sent();
-					SendError::Disconnected
+					SendError::Disconnected(err.0.into())
 				})
 			},
 			_ => Ok(()),
 		}
 	}
 
 	/// Attempt to send message or fail immediately.
-	pub fn try_send(&mut self, msg: T) -> result::Result<(), TrySendError<MaybeTimeOfFlight<T>>> {
+	pub fn try_send(&mut self, msg: T) -> result::Result<(), TrySendError<T>> {
 		let msg = self.prepare_with_tof(msg);
 		self.inner.try_send(msg).map_err(|e| {
-			if e.is_full() {
-				// Count bounded channel sends that block.
-				self.meter.note_blocked();
-			}
 			self.meter.retract_sent();
-			e
+			match e {
+				TrySendError::Full(inner_error) => {
+					self.meter.note_blocked();
```

Comment:
This metric should count the times send blocked, but in this case we do not and return.

Reply:
We should move the metric in the blocking `send` method.

Reply:
That will change the behaviour that is expected in the tests btw (that's why the tests are failing)